src/foam/util/filesystem/FileWalker.js

```diff
@@ -0,0 +1,120 @@
+foam.CLASS({
+  package: 'foam.util.filesystem',
+  name: 'FileWalker',
+  topics: ['files', 'directoryStart', 'directoryEnd', 'error'],
+  documentation: `
+    This FileWalker goes through items in a directory recursively
+    and reports the following to these 4 topics:
+    - files: An object with information about the files in a directory.
+        For example:
+          { directory: 'foam2/src/foam/model',
+            files: {
+              "Model.js": {/* lstat object */} } }
+    - directoryStart: the path of a directory being walked
+    - directoryEnd: the path of a directory done being walked
+    - error: this topic reports two arguments to the listener:
+        the source of the error as a string (ex: 'lstat') and the
+        error object.
+  `,
+  properties: [
+    // TODO: add regexes for ignored files (ex: dotfiles)
+    {
+      name: 'order',
+      value: 'BFS',
+      documentation: `
+        Order of FileWalker; either BFS for breadth-first search,
+        or DFS for depth-first-search.
+
+        Note that for DFS the FileWalker still must stat all the files
+        in a directory before it can move deeper.
+      `
+    },
+    'nodejs_'
+  ],
+  methods: [
+    function init() {
+      this.nodejs_ = {};
+      this.require_('path');
+      this.require_('fs');
+    },
+    function require_(name, asName) {
+      Object.defineProperty(this.nodejs_, asName || name, {
+        value: require(name),
+        writable: false
+      });
+    },
+    function walk(path, depth) {
+      if ( typeof depth === 'undefined' ) {
+        depth = 0;
+      }
+
+      var self = this;
+      return new Promise(function(resolve, reject) {
+        self.nodejs_.fs.readdir(path, function (err, files) {
+          if ( err ) {
+            reject(err);
+            return;
+          }
+
+          filesMessage = {
+            directory: path,
+            files: {}
+          };
+          directoriesToWalk = [];
+
+          files.forEach(function (file) {
+            let fullPath = self.nodejs_.path.join(path, file);
+            let stats = null;
+            try {
+              stats = self.nodejs_.fs.lstatSync(fullPath);
+            } catch (e) {
+              console.warn('Failed to stat file: ' + fullPath);
+              self.error.pub('stat', e);
+              return;
+            }
+            if ( stats.isDirectory() ) {
+              directoriesToWalk.push(fullPath);
+            } else {
+              filesMessage.files[file] = {
+                name: file,
+                fullPath: fullPath,
+                stats: stats,
+              };
+            }
+          });
+
+          let recursePromises = [];
+          switch ( self.order ) {
+          case 'BFS':
+            self.directoryStart.pub(path);
+            self.files.pub(filesMessage);
+            directoriesToWalk.forEach((fullPath) => {
```

Comment:
Another personal preference kind of thing, but any time you create an empty array and loop over something else to populate the new array you can usually just map the old array (and filter if need be (or use `reduce` instead of `map` and `filter` to do both in one go)). Personally I find it easier to read that way since it's more declarative that way.

Eg:

```JavaScript
let ys = [];

for ( let i = 0; i < xs.length; i++ ) {
  ys.push(foo(xs[x]));
}
```

can be written as

```
let ys = xs.map(foo);
```

Eg with filtering:

```JavaScript
let ys = [];

for ( let i = 0; i < xs.length; i++ ) {
  if ( bar(xs[x]) ) {
    ys.push(foo(xs[x]));
  }
}
```

could be written as

```JavaScript
let ys = xs.filter(bar).map(foo);
```

or

```JavaScript
let ys = xs.reduce((acc, x) => bar(x) ? acc.concat([x]) : acc, []);
```